enhancements/sig-policy/89-operator-policy-kind/README.md

```diff
@@ -0,0 +1,557 @@
+manual intervention from an admin.
+
+A new controller can handle the relationships between resources better, for example knowing when an
+OperatorGroup should be created, or identifying the Deployments and CustomResourceDefinitions
+managed by a ClusterServiceVersion. This information could be incorporated into the policy status,
+helping administrators understand the status of the operators from the hub cluster. The new kind can
+also provide new controls over the upgrades and uninstallation of these operators, which would be
+difficult with current solutions.
+
+### Goals
+
+1. It should be easy to install an operator on a managed cluster with just one policy.
+2. It should be possible to specify what upgrades should be allowed performed on an operator.
+3. The policy status should report on the health of the operator, both the OLM resources like the
+   Subscription, as well as the associated Deployments.
+4. The policy should report when the CatalogSource for the operator is unhealthy. It should be
+   configurable whether this will cause the policy to become NonCompliant.
+5. What happens when the policy is deleted should be configurable: the operator might be deleted, or
+   the operator and its CRDs might be deleted, or nothing might be deleted. By default, only the
+   Subscription and CSVs will be deleted (matching the suggestion from OLM).
+
+### Non-Goals
+
+1. The installation of OLM on managed clusters will not be automated by this enhancement.
+2. Management of "Operands" will not be included in this new type - they should be defined in
+   separate `ConfigurationPolicies`, possibly with a
+   [dependency](../74-order-policy-execution/README.md) on the `OperatorPolicy`. ("Operand" here
+   refers to resources that would be created by a user, and managed by the operator, eg instances of
+   CRDs which are defined by the operator.)
+3. Creation of `CatalogSources` will not be included in the new type - they should be defined in
+   separate `ConfigurationPolicies`.
+
+## Proposal
+
+### OperatorPolicy Syntax
+
+```yaml
+apiVersion: policy.open-cluster-management.io/v1beta1
+kind: OperatorPolicy
+metadata:
+  name: my-operator
+spec:
+  remediationAction: enforce # or inform
+  severity: medium
+  operatorGroup: # optional
+    name: my-operator-group
+    namespace: own-namespace
+    target:
+      namespaces:
+        - foo
+        - bar
+    # selector: # One of selector or namespaces is allowed.
+    #   matchLabels: # or matchExpressions
+    #     mydomain.io/prod: "true"
+    serviceAccountName: my-og-account # optional
+  subscription:
+    config: {} # optional
+    channel: stable
+    name: my-operator
+    namespace: own-namespace
+    installPlanApproval: Automatic # may be overridden to Manual by other settings
+    source: my-catalog
+    sourceNamespace: my-catalog-namespace
+    startingCSV: my-operator.v0.1.0 # optional
+  allowedCSVs:
```

Comment:
better name could be **approvedCSV** as allowedCSV meaning more like it is allowed to show up for subscriptions but not approved.

Reply:
I'm thinking about this in the new possible mode of a "mustnothave" OperatorPolicy. In that case, it would be nice to be able to supply a list of specific versions I *don't* want to allow on the managed cluster.

Would a simpler name like `versions` work? Then in a "musthave" OperatorPolicy, this list indicates the versions that can be present for the policy to be compliant. And in "mustnothave" mode, the list would indicate versions that must not be present.

Reply:
okay, make sense for mustnothave case.

Reply:
Are you saying two policies are needed if we want to allow some versions and disallow some others?

Reply:
@jc-rh it depends on if you know the specific versions you want to allow. If you make a "musthave" policy with specific versions, than any versions other than what is specified will make the policy NonCompliant (and in enforce mode, the controller might try to remove that operator).

If you can't restrict the allowed versions to a specific list - maybe you want something like "any version, except x, y, or z" - then in this proposal, two policies would be necessary: a "musthave" with no versions specified, and a "mustnothave" with x, y, and z. (This case wasn't really covered by any previous design iterations.)

I was anticipating "mustnothave" being more useful if the operators aren't necessarily being managed by an OperatorPolicy.

Reply:
Hmm, OK. Our main use case would be `AtLatestKnown`. It would be a pain to update allowed version in the policy for every single update. Occasionally, when they want to disallow a problematic version, it would be awkward to list all currently running/upgradable versions as allowed.